addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/internal/protocol/xml/DeviceInformationXML.java

```diff
@@ -73,46 +76,69 @@ public DeviceInformationXML(AbstractConnection com, DeviceInformationState state
-        state.id = XMLUtils.getNodeContentOrDefault(basicStatus, "System_ID", "");
-        state.version = XMLUtils.getNodeContentOrDefault(basicStatus, "Version", "");
+        state.host = con.getHost();
+        state.name = getNodeContentOrEmpty(systemConfigNode, "Model_Name");
+        state.id = getNodeContentOrEmpty(systemConfigNode, "System_ID");
+        state.version = getNodeContentOrEmpty(systemConfigNode, "Version");
 
         state.zones.clear();
+        state.features.clear();
+        state.properties.clear();
 
-        Node node = XMLUtils.getNode(basicStatus, "Feature_Existence");
-        if (node == null) {
-            throw new ReceivedMessageParseException("Zone information not provided: " + response);
-        }
+        // Get and store the Yamaha Description XML. This will be used to detect proper command naming in other areas.
+        DeviceDescriptorXML descriptor = new DeviceDescriptorXML();
+        descriptor.load(con);
+        descriptor.attach(state);
 
-        for (YamahaReceiverBindingConstants.Zone zone : YamahaReceiverBindingConstants.Zone.values()) {
-            if (isFeatureSupported(node, zone.toString())) {
-                logger.trace("Adding zone: {}", zone);
-                state.zones.add(zone);
+        Node featureNode = getNode(systemConfigNode, "Feature_Existence");
+        if (featureNode != null) {
+            for (Zone zone : Zone.values()) {
+                checkFeature(featureNode, zone.toString(), zone, state.zones);
             }
+
+            XMLConstants.FEATURE_BY_YNC_TAG.forEach((name, feature) -> checkFeature(featureNode, name, feature, state.features));
+
+        } else {
+            // on older models (RX-V3900) the Feature_Existence element does not exist
+
+            descriptor.zones.forEach((zone, x) -> addFeature(zone, state.zones));
+            descriptor.features.forEach((feature, x) -> addFeature(feature, state.features));
         }
 
-        state.supportTuner = isFeatureSupported(node, "Tuner");
-        state.supportDAB = isFeatureSupported(node, "DAB");
-        state.supportSpotify = isFeatureSupported(node, "Spotify");
+        if (state.zones.contains(Zone.Main_Zone) && !state.zones.contains(Zone.Zone_2)) {
+            // Detect if Zone_B is supported (HTR-4069). This will allow Zone_2 to be emulated.
+
+            // Retrieve Main_Zone basic status, from which we will know this AVR supports Zone_B feature.
+            Node basicStatusNode = XMLProtocolService.getZoneResponse(con, Zone.Main_Zone, "<Basic_Status>GetParam</Basic_Status>");
+            String power = getNodeContentOrEmpty(basicStatusNode, "Basic_Status/Power_Control/Zone_B_Power_Info");
+            if (StringUtils.isNotEmpty(power)) {
+                logger.info("Zone_2 emulation enabled via Zone_B");
+                addFeature(Zone.Zone_2, state.zones);
+                addFeature(Feature.ZONE_B, state.features);
+            }
+        }
```

Comment:
23 lines before vs. 35 lines after with comments :)

I will break it apart then.

Reply:
If its easy that would be great, purely looking at line count it looks like 45 maybe that is because of newlines. If there is no easy split just leave it.